forum/qa-plugin/q2a-googleauthenticator-login/q2a-googleauthenticator-layer.php

```diff
@@ -0,0 +1,171 @@
+<?php
+
+require_once GOOGLEAUTHENTICATOR_BASIC_PATH . '/src/GoogleAuthenticator.php';
+
+class qa_html_theme_layer extends qa_html_theme_base
+{
+    public function doctype()
+    {
+        parent::doctype();
+
+        if ('account' === $this->request && true === (bool) qa_opt('googleauthenticator_login')) {
+            $content = [
+                'tags'    => 'method="post" action="' . qa_self_html() . '"',
+                'style'   => 'wide',
+                'title'   => qa_lang('plugin_2fa/title')
+            ];
+
+            $content += $this->enablePlugin();
+
+            $this->content['form_2fa'] = $content;
+            qa_html_theme_base::doctype();
+        }
+    }
+
+    private function getUserQuery($userId): ?array
+    {
+        $users = qa_db_read_all_assoc(
+            qa_db_query_sub(
+                'SELECT us.userid, us.2fa_enabled, us.email, us.handle, us.2fa_change_date, up.points FROM ^users us LEFT JOIN ^userpoints up ON us.userid = up.userid WHERE us.userid=$',
+                $userId['userid'] ?? $userId
+            )
+        );
+
+        return empty($users) ? null : $users[0];
+    }
+
+    private function updateUserEnable2FA($userId, $isEnabled, $secret = null, $recoveryCode = null): ?bool
+    {
+        $time = $this->setTime();
+        $result = qa_db_query_sub(
+            'UPDATE ^users SET 2fa_enabled=#, 2fa_change_date=$, 2fa_secret=$, 2fa_recovery_code=$ WHERE userid=#',
+            $isEnabled,
+            $time,
+            $secret,
+            $recoveryCode,
+            $userId['userid']
+        );
+
+        if (true === $result) {
+            return $isEnabled;
+        }
+
+        user_error(qa_lang('plugin_2fa/2fa_setup_error'));
+    }
+
+    private function enablePlugin(): array
+    {
+        $userAccount   = $this->getUser();
+        [$userActive2fa, $recoveryCode, $secret] = $this->prepareGoogleAuthenticator($userAccount);
+
+        if (!(bool) $userActive2fa) {
+            return $this->render2FADisabledForm();
+        }
+
+        $result = $this->render2FAEnabledForm($userAccount['2fa_change_date']);
+
+        if (isset($this->init)) {
+            $note = qa_lang_html('plugin_2fa/2fa_data_info');
+            $note = str_replace(
+                ['{{ QR_CODE }}', '{{ SECRET }}', '{{ RECOVERY_CODE }}', '{{ ERROR_START }}', '{{ ERROR_END }}'],
+                [
+                    '<br><div style="text-align: center;"><img src="' . $this->init->getQRCode() . '"></div><br>',
```

Comment:
Czy te `<br>`y są potrzebne? Odstęp między elementami można zrobić przy użyciu `margin` w CSS. A skoro o stylowaniu mowa, to przy okazji wypadałoby przenieść ten `text-align: center;` do CSSa.

Brakuje atrybutu `[alt]` dla obrazka - mógłby mieć wartość np. "QRCode".

Reply:
Jeśli chcesz, możesz wprowadzić zmiaany - nie mam zbyt wiele kompetencji do edycji html/css.